staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -59,3 +63,89 @@ type InTreePlugin interface {
 	// RepairVolumeHandle generates a correct volume handle based on node ID information.
 	RepairVolumeHandle(volumeHandle, nodeID string) (string, error)
 }
+
+// replaceTopology overwrites an existing topology by a new one.
+func replaceTopology(pv *v1.PersistentVolume, oldKey string, newTopology v1.NodeSelectorRequirement) error {
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		for j, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
+			if r.Key == oldKey {
+				pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions[j] = newTopology
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("could not find topology key %q", oldKey)
+}
+
+// getTopology returns the current topology zones with the given key.
+func getTopology(pv *v1.PersistentVolume, key string) []string {
+	if pv.Spec.NodeAffinity == nil ||
+		pv.Spec.NodeAffinity.Required == nil ||
+		len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) < 1 {
+		return nil
+	}
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		for _, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
+			if r.Key == key {
+				return r.Values
```

Comment:
I think you need to continue iterating through all the terms, because you could technically express Regional PD topology two ways:

1) With 1 node selector term, and 2 values in the expression. ie (OR (AND (OR zone1, zone2)))
2) With 2 node selector terms, and 1 value each expression. ie (OR (AND (OR zone1)), (AND (OR zone2))

Reply:
You can take a look at https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/topology.go#L395 for an explanation and better example. Although I don't think we need to have the function here be as generic, since we know the exact structure of the cloud volumes.

Reply:
Please see if my changes make sense. I also changed the tests to reflect that.